### PR TITLE
Add LFSC proofs for INSTANTIATE and BETA_REDUCE

### PR DESCRIPTION
**Pull request by ajreynol**
Also updates the LFSC signatures and LFSC to use `function` instead of `program` to ensure the results of side conditions are locally cached.

### CODE REVIEW

**Review by abdoo8080 — CHANGES_REQUESTED** [review]
The PR looks good overall. Are the changes here already tested by the current set of benchmarks?

File: proofs/lfsc/signatures/quantifiers_rules.plf [hunk 1]
```diff
@@ -47,3 +47,29 @@
...
 (declare skolem_intro (! t term
                         (holds (= (skolem t) t))))
+
+; Beta-reduce t
+; We expect t to be of the form
+;   (apply (lambda v1 s1
+;     (apply (lambda v2 s2
+;     ...
+;     t2))
+;   t1))
```
Comment (abdoo8080) [inline proofs/lfsc/signatures/quantifiers_rules.plf:57]: The locations of some closing parentheses look off to me. Did you mean?
```suggestion
;   (apply (lambda v1 s1
;     (apply (lambda v2 s2
;     ...)
;     t2))
;   t1)
```

**Review by abdoo8080 — CHANGES_REQUESTED** [review]

File: src/proof/lfsc/lfsc_post_processor.cpp [hunk 2]
```diff
@@ -396,6 +396,44 @@ bool LfscProofPostprocessCallback::update(Node res,
...
     }
     break;
+    case PfRule::INSTANTIATE:
+    {
+      Node q = children[0];
+      Assert(q.getKind() == FORALL);
+      std::vector<Node> terms;
+      std::vector<Node> qvars(q[0].begin(), q[0].end());
+      Node conc = q;
+      for (size_t i = 0, nvars = q[0].getNumChildren(); i < nvars; i++)
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/proof/lfsc/lfsc_post_processor.cpp:406]: No, the INSTANTIATE proof rule only is for full instantiation.

File: proofs/lfsc/signatures/quantifiers_rules.plf (hunk 1 above)
Comment (ajreynol) [reply proofs/lfsc/signatures/quantifiers_rules.plf:57]: I find the updated version more confusing.  Note there are 2 opening parens on line 53, they are both closed on line 57.

Comment (ajreynol) [reply proofs/lfsc/signatures/quantifiers_rules.plf:57]: Nvm, the nesting was indeed wrong, I'll update.

**Review by abdoo8080 — APPROVED** [review]

**Pull request merged**